File: 2023_january/min_time_collect_all_apples_in_a_tree/main.go
package main

import "fmt"

func minTime(n int, edges [][]int, hasApple []bool) int {
	var steps int

	return steps
}

func main() {
	n, edges, hasApple, exp := 7, [][]int{{0, 1}, {0, 2}, {1, 4}, {1, 5}, {2, 3}, {2, 6}}, []bool{false, false, true, false, true, true, false}, 8
	// n, edges, hasApple, exp := 7, [][]int{{0, 1}, {0, 2}, {1, 4}, {1, 5}, {2, 3}, {2, 6}}, []bool{false, false, true, false, false, true, false}, 6
	// n, edges, hasApple, exp := 7, [][]int{{0, 1}, {0, 2}, {1, 4}, {1, 5}, {2, 3}, {2, 6}}, []bool{false, false, false, false, false, false, false}, 0
	res := minTime(n, edges, hasApple)
	fmt.Printf("expected %v, result %v", exp, res)
}

File: 2023_january/min_time_collect_all_apples_in_a_tree/main.py
from collections import defaultdict
from typing import List

class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        maps = defaultdict(list)
        for x, y in edges:
            maps[x].append(y)
            maps[y].append(x)


        def traverse(node, parent):
            res = 0
            for child in maps[node]:
                if child != parent:
                    res += traverse(child, node)
            if res or hasApple[node]:
                return res + 2
            return res
        
        return max(traverse(0, -1)-2, 0)

if __name__ == "__main__":
    # n = 7
    # edges = [[0,1],[0,2],[1,4],[1,5],[2,3],[2,6]]
    # hasApple = [False,False,True,False,True,True,False]

    
    n = 7
    edges = [[0,1],[0,2],[1,4],[1,5],[2,3],[2,6]]
    hasApple = [False,False,True,False,False,True,False] 
    # hasApple = [False,False,False,False,False,False,False]
    s = Solution()
    res = s.minTime(n, edges, hasApple)
    print(res)